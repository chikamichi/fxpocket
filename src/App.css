@-webkit-keyframes sk-scaleout {
  0% { -webkit-transform: scale(0) }
  100% {
    -webkit-transform: scale(1.0);
    opacity: 0;
  }
}

@keyframes sk-scaleout {
  0% {
    -webkit-transform: scale(0);
    transform: scale(0);
  } 100% {
    -webkit-transform: scale(1.0);
    transform: scale(1.0);
    opacity: 0;
  }
}

body {
  background: cornsilk;
}

#root {
  display: flex;
  justify-content: center;
}

.fxp {
  text-align: center;
}

.fxp-title {
  margin-bottom: 0;
  color: brown;
}

.fxp-motto {
  margin-top: 0;
  font-size: 1em;
  color: slategray;
}

/* fxPocket App => fxp- namespace */
.fxp-app {
  display: flex;
}

/* Borrowed from http://tobiasahlin.com/spinkit/ */
.fxp-loading {
  width: 40px;
  height: 40px;
  margin: 100px auto;
  background-color: #333;
  border-radius: 100%;
  -webkit-animation: sk-scaleout 1.0s infinite ease-in-out;
  animation: sk-scaleout 1.0s infinite ease-in-out;
}

.fxp-currency {
  display: flex;
  flex-grow: 1;
  padding: 20px;
}

.fxp-currency--quote {
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
}

.fxp-currency--counter {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
}

.fxp-currency__result {
  margin-top: 0.8em;
}

.fxp-currency__label {
  padding-left: 0.8em;
}
