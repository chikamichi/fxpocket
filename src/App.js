import React, { Component } from 'react'
import 'whatwg-fetch'

import Currency from './Currency'
import { round } from './utils'
import Fixer from './fixer'
import './App.css'

const INITIAL_STATE = {
  // Keeps track of app's startup process.
  init: false,
  // A list of currency widgets to display at startup.
  widgets: ['EUR', 'USD'],
  // Keeps track of user intents.
  base: {
    amount: undefined,
    currency: undefined
  }
}

class App extends Component {
  constructor(props) {
    super(props)
    this.state = INITIAL_STATE
  }

  onAmountEdited(payload) {
    this.setState({base: payload})
  }

  onCurrencyEdited(payload) {
    const idx = payload.uuid
    const newWidgetsState = [
      ...this.state.widgets.slice(0, idx),
      payload.currency,
      ...this.state.widgets.slice(idx+1)
    ]
    const newBaseState = {
      amount: payload.amount,
      currency: payload.currency
    }
    this.setState({
      widgets: newWidgetsState,
      base: newBaseState
    })
  }

  convertAmounts() {
    const base = this.state.base
    if (!base.amount || !base.currency) return {}
    return this.fixer.currencies.reduce((acc, currency) => {
      let amount
      if (base.currency === currency)
        amount = base.amount
      else
        amount = base.amount * this.fixer.rates[base.currency][currency]
      acc[currency] = round(amount, 2)
      return acc
    }, {})
  }

  componentDidMount() {
    Fixer.connect().then(fixer => {
      this.fixer = fixer
      this.setState({init: true})
    })
  }

  render() {
    if (!this.state.init) return <p>loading app</p>
    const currencies = this.fixer.currencies
    const cb = {
      onAmountEdited: this.onAmountEdited.bind(this),
      onCurrencyEdited: this.onCurrencyEdited.bind(this)
    }
    const amounts = this.convertAmounts()
    const widgets = this.state.widgets.map((currency, i) => {
      return <Currency
        key={i} // key is simply the position in this.state.widgets
        uuid={i}
        type='quote' // TODO: remove this type prop
        amount={amounts[currency]}
        currency={currency}
        currencies={currencies}
        onAmountEdited={cb.onAmountEdited}
        onCurrencyEdited={cb.onCurrencyEdited} />
    })
    return (
      <div className="fxp">
        {widgets}
      </div>
    )
  }
}

export default App
