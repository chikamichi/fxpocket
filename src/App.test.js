import React from 'react';
import ReactDOM from 'react-dom'
import { shallow } from 'enzyme'

import App from './App'

describe('<App />', () => {
  describe('layout', () => {
    var wrapper

    beforeEach(async () => {
      wrapper = await shallow(<App />)
      wrapper.update()
    })

    it('looks like the registered snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('Currency#convertAmounts()', () => {
    var wrapper

    beforeEach(async () => {
      wrapper = await shallow(<App />)
      wrapper.update()
    })

    it('returns a null object when state.base is not set', () => {
      expect(wrapper.instance().convertAmounts()).toEqual({})
    })

    it('computes each widget\'s amount based on state.base insights', () => {
      wrapper.setState({base: {amount: 2, currency: 'EUR'}})
      expect(wrapper.instance().convertAmounts()).toEqual({
        EUR: 2, // 2*1
        USD: 2.5 // 2*1.25
      })
      wrapper.setState({base: {amount: 5, currency: 'USD'}})
      expect(wrapper.instance().convertAmounts()).toEqual({
        EUR: 4, // 5*0.8
        USD: 5 // 5*1
      })
    })

    it('truncates float to a precision of 2 decimals', () => {
      // Same precision fixer.io exposes its rates and people usually expect
      // amounts to be expressed as.
      wrapper.setState({base: {amount: 1.239, currency: 'EUR'}})
      expect(wrapper.instance().convertAmounts()).toEqual({
        EUR: 1.24, // 1.239*1 truncated
        USD: 1.55 // 1.239*1.25 = 1.54875 truncated
      })
    })
  })

  describe('state', () => {
    var wrapper

    beforeEach(async () => {
      wrapper = await shallow(<App />)
      wrapper.update()
    })

    // describe('.currencies', () => {
    //   it('is stored as a [{currency => rates}+] object', () => {
    //     const expected = Object.keys(responses.fixer.latest).reduce((acc, currency) => {
    //       acc[currency] = responses.fixer.latest[currency].rates
    //       return acc
    //     }, {})
    //     expect(wrapper).toHaveState('currencies', expected)
    //   })
    // })

    describe('.widgets', () => {
      const initialState = ['EUR', 'USD']

      it('initially instructs to display two widgets, EUR and USD', () => {
        expect(wrapper).toHaveState('widgets', initialState)
      })

      it('is updated upon onCurrencyEdited()', () => {
        const payload = {
          uuid: 0,
          amount: 42,
          currency: 'USD'
        }
        const newState = ['USD', 'USD']
        wrapper.instance().onCurrencyEdited(payload)
        expect(wrapper).toHaveState('widgets', newState)
      })
    })

    describe('.base', () => {
      const initialState = {
        amount: undefined,
        currency: undefined
      }

      it('is initially empty', () => {
        expect(wrapper).toHaveState('base', initialState)
      })

      it('is updated upon onAmountEdited()', () => {
        const newState = {
          amount: 42,
          currency: 'EUR'
        }
        wrapper.instance().onAmountEdited(newState)
        expect(wrapper).toHaveState('base', newState)
      })

      it('is updated upon onCurrencyEdited()', () => {
        const payload = {
          uuid: 0,
          amount: 42,
          currency: 'USD'
        }
        const newState = {
          amount: 42,
          currency: 'USD'
        }
        wrapper.instance().onCurrencyEdited(payload)
        expect(wrapper).toHaveState('base', newState)
      })
    })
  })
})
