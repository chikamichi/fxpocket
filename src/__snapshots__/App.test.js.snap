// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> layout looks like the registered snapshot 1`] = `
<div
  className="fxp"
>
  <Currency
    currencies={
      Array [
        "EUR",
        "USD",
      ]
    }
    currency="EUR"
    key="0"
    onAmountEdited={[Function]}
    onCurrencyEdited={[Function]}
    type="quote"
    uuid={0}
  />
  <Currency
    currencies={
      Array [
        "EUR",
        "USD",
      ]
    }
    currency="USD"
    key="1"
    onAmountEdited={[Function]}
    onCurrencyEdited={[Function]}
    type="quote"
    uuid={1}
  />
</div>
`;
